trecs/components/users.py

```diff
@@ -189,19 +327,19 @@ def __init__(
         if actual_user_profiles is None and not isinstance(size, tuple):
             raise TypeError("size must be a tuple, is %s" % type(size))
         if actual_user_scores is not None:
-            if not isinstance(actual_user_scores, (list, np.ndarray)):
+            if not isinstance(actual_user_scores, (list, np.ndarray, sp.spmatrix)):
                 raise TypeError("actual_user_profiles must be a list or numpy.ndarray")
         if actual_user_profiles is None and size is not None:
             row_zeros = np.zeros(size[1])  # one row vector of zeroes
-            while actual_user_profiles is None or contains_row(actual_user_profiles, row_zeros):
+            while actual_user_profiles is None or mo.contains_row(actual_user_profiles, row_zeros):
                 # generate matrix until no row is the zero vector
                 actual_user_profiles = self.rng.normal(size=size)
```

Comment:
Can we generate a sparse matrix by default? I don't expect people to actually use the defaults unless they're learning how to use each component, so this would be just an incentive for them to learn the "fast" way from the outset

Reply:
Hmm, we could generate a sparse matrix by default, but whether or not sparse matrices are actually "faster" depends on the sparsity of the data, right? By default, we have sparse 0/1 item attributes and user profiles that are normally distributed, so it makes sense for item attributes to be sparse and for user profiles to be dense. That means the user-item score matrix will be dense (because an arbitrary user-item score is not likely to be exactly zero). So I'm not 100% sure that sparse matrices are actually fast right now based on our default assumptions, I am 100% sure that they allow us to run Bass simulations without running out of memory...

Reply:
In a "realistic" setting, I'd expect item attributes to be very sparse (an item is likely to belong only to a few categories, not the majority of them) and user-item scores to be on average quite sparse too, even if user profiles are dense, because of the sparsity of item attributes. Is this belief correct? If so, what, if anything, needs to change to address it?

Reply:
I'm not sure if that lines up with my understanding - if user profiles are dense and item attributes are very sparse, then every user-item pair will be most likely nonzero, because if user = (u1 u2 u3 u4 .. u_n) and item = (0 0 0 0 .. 1), <user, item> = u_n. So typically every user will have a nonzero score associated with every item, thus making the user-item score matrix dense, not sparse. Does that make sense?

Reply:
Yes, that makes sense. I agree with leaving the defaults dense. Let me explain what I thought so we can make sure we agree.

I got confused for a second -- I guess my question was actually: does it make sense to expect the model to have user preferences, of the form `user = (u1 u2 u3 u4 .. u_n)`, that are not sparse. I think based on that, my reasoning for user-item scores to be sparse only applies to specific models (e.g., content filtering) where it's reasonable to expect user profiles to be sparse as well.

If all of this doesn't raise any other question, then let's leave the defaults the way they are. We can think about whether it'll make sense to change the defaults on a case by case basis, which is definitely not a priority now.

Reply:
Sounds good to me - I agree w/ everything that you said!